lib/recursive-methods.rb

```diff
@@ -1,49 +1,80 @@
-# Authoring recursive algorithms. Add comments including time and space complexity for each method.
-
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because it makes as many calls as the number is large
+# Space complexity: O(n) because the stack holds as many calls as the number is large
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return 1 if n == 1 || n == 0
+  raise ArgumentError if n < 0
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), 'n' being the length of the string
+# Space complexity: O(n) because the stack holds as many calls as the string is long
```

Comment:
Because `s[0..-2]` creates a new array and copies the entries over from the old array... This is O(n<sup>2</sup>) for both time and space complexity!